US1009117: Remove reliance on large messages being supported

DavidGerald-Milligan opened the pull request:
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=1009117

DavidGerald-Milligan committed:
WIP

DavidGerald-Milligan committed:
Store and retreive

DavidGerald-Milligan committed:
Reviewing

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](https://sou-jenkins2.swinfra.net/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=WorkerFramework&REPO_NAME=worker-framework&BRANCH_NAME=US1009117) to go ahead and create the build...

DavidGerald-Milligan committed:
Use headers

DavidGerald-Milligan committed:
Reviewed

DavidGerald-Milligan committed:
restored ws

DavidGerald-Milligan committed:
Test update

DavidGerald-Milligan committed:
Test update

DavidGerald-Milligan committed:
readme update

DavidGerald-Milligan committed:
confirm listener update

DavidGerald-Milligan committed:
confirm listener update

DavidGerald-Milligan committed:
confirm listener update

DavidGerald-Milligan committed:
publisher update

DavidGerald-Milligan committed:
publisher update

DavidGerald-Milligan committed:
Add header handling to tests

DavidGerald-Milligan committed:
removed optional nullable

DavidGerald-Milligan committed:
restored ws

DavidGerald-Milligan committed:
Relocated message cfg class

DavidGerald-Milligan committed:
REnamed message cfg

DavidGerald-Milligan committed:
restored ws

DavidGerald-Milligan committed:
fixed renaming issue

DavidGerald-Milligan committed:
Moved rehydration into the rabbit consumer

DavidGerald-Milligan committed:
Test updates

DavidGerald-Milligan committed:
Minor refactoring

DavidGerald-Milligan committed:
Refactoring etc

buildmachine-sou-jenkins2 commented:
CI Build Link:
https://sou-jenkins2.swinfra.net/job/WorkerFramework/job/WorkerFramework~worker-framework~US1009117~CI

buildmachine-sou-jenkins2 commented:
The Documentation QA site for this branch has been built:
https://workerframework-ci-worker-framework-us1009117-5a4fcb.glpages.otxlab.net

DavidGerald-Milligan committed:
WIP

DavidGerald-Milligan committed:
Do not add invalid messageid to task information

DavidGerald-Milligan committed:
Merge branch 'US1009117' into US1009117-IT

DavidGerald-Milligan committed:
issue with read only headers, and IT

DavidGerald-Milligan committed:
Merge branch 'develop' into US1009117

DavidGerald-Milligan committed:
Map update

DavidGerald-Milligan committed:
Issues resolved

DavidGerald-Milligan committed:
Issues resolved

DavidGerald-Milligan committed:
Merge branch 'US1009117-IT' into US1009117

DavidGerald-Milligan committed:
Merge branch 'develop' into US1009117

DavidGerald-Milligan committed:
Refactored previous header removal

DavidGerald-Milligan committed:
Use env var for webdav url

DavidGerald-Milligan committed:
Remove unused imports

DavidGerald-Milligan committed:
updated javadoc

DavidGerald-Milligan commented on worker-core/src/main/java/com/github/workerframework/core/WorkerCore.java:
```diff
@@ -151,12 +151,11 @@ public CoreTaskCallback(final Codec codec, final WorkerStats stats, final Worker
         {
             Objects.requireNonNull(taskInformation);
             stats.incrementTasksReceived();
-            stats.getInputSizes().update(taskMessage.length);
```
Discussed and identified as unused, removed to save deserializing the taskmessage again.

andyreidz replied:
Remove from WorkerApplication.

        metrics.register(MetricRegistry.name("core.inputSizes"), core.getStats().getInputSizes());

and from WorkerStats

DavidGerald-Milligan commented on worker-core/src/main/java/com/github/workerframework/core/WorkerCore.java:
```diff
@@ -219,9 +216,7 @@ private void registerNewTaskImpl(final TaskInformation taskInformation, final by
+            } catch (final InvalidJobTaskIdException ijte) {
                 throw new InvalidTaskException("TaskMessage contains an invalid job task identifier", ijte);
             }
         }
```
Desialization now done earlier in WorkerQueueConsumerImpl.java

DavidGerald-Milligan committed:
Removed unused test dependency

DavidGerald-Milligan committed:
Return exception from deserializeTaskMessage

DavidGerald-Milligan committed:
Refactored WorkerCoreTest

DavidGerald-Milligan committed:
update defaut config readme

DavidGerald-Milligan committed:
Update worker-queue-rabbit readme

DavidGerald-Milligan committed:
Removed ws

DavidGerald-Milligan committed:
Refactored RabbitWorkerQueueConfig

DavidGerald-Milligan committed:
Remove redundant null check

DavidGerald-Milligan committed:
Reverted map

DavidGerald-Milligan committed:
Update publisher

DavidGerald-Milligan committed:
Update consumer

DavidGerald-Milligan committed:
Update consumer

DavidGerald-Milligan committed:
Update consumer test class

DavidGerald-Milligan committed:
Update pulisher test class

DavidGerald-Milligan committed:
Update POM versions to 10.0.0-SNAPSHOT

DavidGerald-Milligan committed:
Update release notes

andyreidz commented on util-rabbitmq/src/main/java/com/github/workerframework/util/rabbitmq/RabbitHeaders.java:
```diff
@@ -23,4 +23,5 @@ public class RabbitHeaders
     public static final String RABBIT_HEADER_CAF_WORKER_REJECTED = "x-caf-worker-rejected";
     public static final String RABBIT_HEADER_CAF_WORKER_RETRY = "x-caf-worker-retry";
     public static final String RABBIT_HEADER_CAF_DELIVERY_COUNT = "x-delivery-count";
+    public static final String RABBIT_HEADER_CAF_DEHYDRATION_ID = "x-dehydration-id";
```
```suggestion
    public static final String RABBIT_HEADER_CAF_DEHYDRATION_ID = "x-caf-dehydration-id";
```

andyreidz commented on worker-core/src/main/java/com/github/workerframework/core/WorkerCore.java:
```diff
@@ -151,12 +151,11 @@ public CoreTaskCallback(final Codec codec, final WorkerStats stats, final Worker
          */
         @Override
-        public void registerNewTask(final TaskInformation taskInformation, final byte[] taskMessage, Map<String, Object> headers)
+        public void registerNewTask(final TaskInformation taskInformation, final TaskMessage taskMessage, Map<String, Object> headers)
```
import com.github.cafapi.common.api.DecodeMethod; is unused

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitTaskInformation.java:
```diff
@@ -30,19 +31,47 @@ public class RabbitTaskInformation implements TaskInformation {
+        this(inboundMessageId, isPoison, Optional.empty());
+    }
+
+    public RabbitTaskInformation(final String inboundMessageId, final boolean isPoison, final Optional<String> dehydratedMessageId) {
```
```suggestion
    public RabbitTaskInformation(final String inboundMessageId, final boolean isPoison, final Optional<String> dehydratedTaskMessageStorageRef) {
```

DavidGerald-Milligan committed:
Conflicts resolved

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java:
```diff
@@ -66,18 +86,47 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
+    ) throws QueueException {
+        try {
+            if (shouldStoreTaskMessage(taskMessage.length)) {
+                final TaskMessage outboundTaskMessage = codec.deserialise(taskMessage, TaskMessage.class);
```
Change handlePublish to accept TaskMessage instead of byte[] so we avoid deserialization here.

DavidGerald-Milligan replied:
@andyreidz I see why you say this but this creates an issue in the consumer if we cannot deserialize the incoming byte array to a TaskMessage and have to republish.  Is it a valid option to reject the delivery at this point as we do in handlePublish if it cannot be deserialized? https://github.com/WorkerFramework/worker-framework/blob/US1009117/worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java#L135

DavidGerald-Milligan replied:
actually, we have to either deserialize or serialize here anyway, both storing and publishing require one or other of the actions.

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -180,14 +227,17 @@ private void processReject(long id, boolean requeue)
         headers.put(RabbitHeaders.RABBIT_HEADER_CAF_WORKER_RETRY, String.valueOf(retries + 1));
+        if (dehydratedMessageId.isPresent()) {
+            headers.put(RABBIT_HEADER_CAF_DEHYDRATION_ID, dehydratedMessageId.get());
+        }
```
For a poison message, using classic queues, won't the dehydratedMessageId be deleted when the delivered message is ack'd?

Also the delivery.getMessageData() will return an empty byte array for a dehydrated message?

DavidGerald-Milligan replied:
_If we call republishClassicRedelivery, the incoming dehydratedMessageId is not set in the TaskInformation so it will not be deleted.   Yes the messageData will be empty if its a dehydrated message, but it will be re-hydrated on re-delivery._

**Update**: I didn't consider the message size on republish would mean the header is not set,  we need to read the stored message here pass that on for republishing and then delete the original message here, or find a way to pass the existing reference to the publisher.

andyreidz reviewed (changes requested):
Main thing is poison message handling which will need testing for both quorum and classic queue modes, it's likely that classic is problem area because it publishes a new message which is a copy of the original with a modified header.

DavidGerald-Milligan committed:
REmove deleted file

DavidGerald-Milligan committed:
Update util-rabbitmq/src/main/java/com/github/workerframework/util/rabbitmq/RabbitHeaders.java

Co-authored-by: andyreidz <andrew.reid@microfocus.com>

DavidGerald-Milligan committed:
Updated dehydration header name

DavidGerald-Milligan committed:
Updated dehydration cfg var names in readme

DavidGerald-Milligan committed:
Refactored to taskMessageStorageRef

DavidGerald-Milligan committed:
removed unused import

DavidGerald-Milligan committed:
removed core.inputSizes

DavidGerald-Milligan committed:
Removed deserializeation/serialization from getOutboundByteArray

DavidGerald-Milligan committed:
Update rabbit module unit tests

DavidGerald-Milligan committed:
Reviewed rejected message handling

DavidGerald-Milligan committed:
minot refactoring

DavidGerald-Milligan committed:
minor refactoring

DavidGerald-Milligan committed:
minor refactoring

DavidGerald-Milligan committed:
Test updates

DavidGerald-Milligan committed:
Deleted datastore files

DavidGerald-Milligan committed:
Rename dehydrate

DavidGerald-Milligan committed:
Corrected pom

DavidGerald-Milligan committed:
Undoing copilot errors

DavidGerald-Milligan committed:
corrected readme

DavidGerald-Milligan committed:
Undoing copilot errors

DavidGerald-Milligan committed:
Fixed failing IT

DavidGerald-Milligan committed:
Fixing tests

DavidGerald-Milligan committed:
Removed codec in test ctor

DavidGerald-Milligan committed:
Renaming dehydrate

DavidGerald-Milligan committed:
Merge conflicts

DavidGerald-Milligan committed:
Renaming

DavidGerald-Milligan committed:
renaming

DavidGerald-Milligan committed:
Reverted PoisonMessageIT

DavidGerald-Milligan committed:
Skip surefire in worker-test

DavidGerald-Milligan committed:
PoisonMessageIT read from datastore

DavidGerald-Milligan committed:
Updated GetWorkerNameIT

DavidGerald-Milligan committed:
commented

DavidGerald-Milligan committed:
Renaming to distinguish between stored message refs

DavidGerald-Milligan committed:
Reversed checks in eorker name test

DavidGerald-Milligan committed:
Test updates

DavidGerald-Milligan committed:
Trying to get one of the tests to pass

DavidGerald-Milligan committed:
Declaring worker in

DavidGerald-Milligan committed:
RE-enable PoisonMessageIT

DavidGerald-Milligan committed:
Updated PoisonMessageIT

DavidGerald-Milligan committed:
Ignore GetWorkerNameIT

DavidGerald-Milligan committed:
Ignoreing flakey tests

DavidGerald-Milligan committed:
Added new poison message test

DavidGerald-Milligan committed:
Added check for poison error message

DavidGerald-Milligan committed:
Added test failure message

DavidGerald-Milligan committed:
Added test failure message

DavidGerald-Milligan committed:
Reverted failing tests, and ignored

DavidGerald-Milligan committed:
Defaulted trackingInfo jobTaskId

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -100,20 +118,72 @@ public void processDelivery(Delivery delivery)
+                String.valueOf(inboundMessageId),
+                isPoison,
+                taskMessageStorageRefOpt,
+                Optional.of(String.format("%s/%s", routingKey, trackingJobTaskId))
```
Just store the trackingJobTaskId here.

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java:
```diff
@@ -66,18 +82,52 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
+                // The stored message will be deleted in the confirm listener
+                publishHeaders.remove(RABBIT_HEADER_CAF_MINIMIZATION_ID);
+            }
+            final var outboundByteArray = getOutboundByteArray(data, taskInformation.getTaskMessagePartialRef(), publishHeaders);
```
taskInformation.getTaskMessagePartialRef is the queue the message was delivered to.

When publishing, the partial refererence should be constructed from the routingKey passed to handlePublish so we store the TaskMessage in a location associated with the publish queue, not the delivery queue.

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitTaskInformation.java:
```diff
@@ -30,19 +31,56 @@ public class RabbitTaskInformation implements TaskInformation {
     private static final Logger LOG = LoggerFactory.getLogger(RabbitTaskInformation.class);
     private final boolean isPoison;
+    private final Optional<String> minimizedTaskMessageStorageRef;
+    private final Optional<String> taskMessagePartialRef;
```
Sorry, I see now this was a bad idea.

Look in the WorkerPublisherImpl.handlePublish, there we're using this to store the TaskMessage bytes in the Datastore when we should be storing to the routingKey where we intend to send the message to.

Change taskMessagePartialRef to trackingInfoJobTaskId

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java:
```diff
@@ -66,18 +82,52 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
             LOG.debug("Publishing message to {} with ack id {}", routingKey, taskInformation.getInboundMessageId());
+            final var publishHeaders = new HashMap<>(headers);
+            // Remove any previous minimization id
+            final Optional<String> inboundTaskMessageStorageRef = Optional.ofNullable(
```
Are there any downsides to simplifying this to always remove the RABBIT_HEADER_CAF_MINIMIZATION_ID header from publishHeaders?

DavidGerald-Milligan replied:
always removing this now

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java:
```diff
@@ -66,18 +82,52 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
+            }
+            return taskMessage;
+        } catch (final Exception e) {
+            throw new QueueException("Error minimizing task message", e);
```
Why are we catching Exception and mapping to QueueException?

Only DataStoreException can be thrown from that block, so it seems like we should let that propagate out to handlePublish and handle it there.

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -177,17 +247,20 @@ private void processReject(long id, boolean requeue)
+     * @param taskMessageStorageRefOpt
      */
-    private void republishClassicRedelivery(final Delivery delivery, final int retries) {
+    private void republishClassicRedelivery(final Delivery delivery, final int retries, final Optional<String> taskMessageStorageRefOpt) {
```
We're reusing the storage reference from the delivered message when publishing to the configured retry queue. 

How are we preventing the WorkerConfirmListener.handleAck from deleting this storage reference when the message is published to the retry queue?

Additionally, the retry queue is usually configured to be the same as the input queue but it's possible that it could be a different queue entirely. 

If that is the case the message sent to that retry queue would have a header with a storage reference containing the original queue's name and not the name of the retry queue.

It's better to avoid reusing the storage reference from the delivered message just create a new copy using retryRoutingKey in the partial reference, then we don't have to worry about the confirm listener.

There's a comment in the ticket saying that we ask operators to purge reject queues. With this new functionality that process will also involve them deleting the location in storage associated with that queue.  That's why it's important that the storage reference in the header always aligns with the queue the message is sitting on.

We also need to think about how this works will message prioritization as the queue names with be staging queues, and it's message prioritization worker that moves them to the intended queue.

DavidGerald-Milligan replied:
I think we are done here

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -100,20 +118,72 @@ public void processDelivery(Delivery delivery)
+                        outputStream.write(buffer, 0, length);
+                    }
+                }
+                return codec.deserialise(outputStream.toByteArray(), TaskMessage.class, DecodeMethod.LENIENT);
```
```suggestion
               return codec.deserialise(deliveryMessageData, TaskMessage.class, DecodeMethod.LENIENT);
```

andyreidz replied:
It might be better to capture the byte array into a variable that you can reuse it in the catch blocks when sending the WorkerPublishQueueEvents

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -100,20 +118,72 @@ public void processDelivery(Delivery delivery)
+                return codec.deserialise(outputStream.toByteArray(), TaskMessage.class, DecodeMethod.LENIENT);
+            }
+           return codec.deserialise(deliveryMessageData, TaskMessage.class, DecodeMethod.LENIENT);
+        } catch (final IOException | CodecException | DataStoreException e) {
```
If you apply the above suggestion then IOException can be removed.

```suggestion
        } catch (final CodecException | DataStoreException e) {
```

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerConfirmListener.java:
```diff
@@ -123,4 +131,13 @@ private void handle(long sequenceNo, boolean multiple, Function<RabbitTaskInform
+            LOG.error("Failed to delete a minimized message:{} from the datastore", taskMessageStorageRef, e);
+        }
+    }
+}
```
newline

DavidGerald-Milligan replied:
done

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -100,20 +118,72 @@ public void processDelivery(Delivery delivery)
+            callback.registerNewTask(taskInformation, taskMessage, delivery.getHeaders());
         } catch (InvalidTaskException e) {
-            LOG.error("Cannot register new message, rejecting {}", taskInformation.getInboundMessageId(), e);
+            final RabbitTaskInformation taskInformation = new RabbitTaskInformation(
```
I made comments on republishClassicRedelivery about the dangers of reusing the storage reference when republishing a new message. If we stop doing that then we can these catch blocks can be simplified to the original approach with the exception of the required code to create a copy of the data in the datastore.

We should handle a problem retrieving from the datastore by adding a entirely new header to record the original storage reference and details of the exception.

andyreidz reviewed (changes requested):
Let's talk when you've have a chance to go through my comments.

DavidGerald-Milligan committed:
Only store trackingJobTaskId in TaskInformation

DavidGerald-Milligan committed:
Remove catch rethrow in getOutboundByteArray

DavidGerald-Milligan committed:
Some review comments addressed

DavidGerald-Milligan committed:
newline

DavidGerald-Milligan commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java:
```diff
@@ -66,18 +82,40 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
+            return new byte[0];
+        }
+        return taskMessage;
+    }
```
note to self: republishClassicRedelivery does not supply trackingJobTaskId

DavidGerald-Milligan committed:
Rework to allow republishClassicRedelivery to function

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -29,67 +36,133 @@
+            try {
+                taskMessageData = retrieveFromDatastore(taskMessageStorageRefOpt.get());
+            } catch (final IOException | DataStoreException e) {
+                // The message was minimized, but we could not retrieve it from the data store.
```
Extract this catch block and the one starting on ln139 into a method, they seem to be the same.

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -100,21 +173,62 @@ public void processDelivery(Delivery delivery)
+        }
+    }
+
+    private byte[] retrieveFromDatastore(final String taskMessageStorageRef) throws DataStoreException, IOException {
```
I don't think we need this buffering.

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -29,67 +36,133 @@
+        // If the message is minimized, we need to retrieve it from the data store.
+        final byte[] taskMessageData;
+        if (taskMessageStorageRefOpt.isPresent()) {
+            try {
```
```suggestion
            try (final var inputStream = dataStore.retrieve(taskMessageStorageRefOpt.get())) {
                taskMessageData = inputStream.readAllBytes();
```

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -29,67 +36,133 @@
+        final byte[] taskMessageData;
+        if (taskMessageStorageRefOpt.isPresent()) {
+            try {
+                taskMessageData = retrieveFromDatastore(taskMessageStorageRefOpt.get());
```
```suggestion
```

andyreidz committed:
1009117 Refactored for readability (#225)

* Update WorkerQueueConsumerImpl.java

* Rename from minimization to payload offload

* Reuse existing headers

* Update WorkerPublisherImpl.java

* Update RabbitWorkerQueueConsumerTest.java

* Some missed refactoring name

* Fix for poison message handling

* Update WorkerQueueConsumerImpl.java

* Handle dead end workers

* Test updates

* Improve deletion of offloaded payloads

* Update WorkerQueueConsumerImpl.java

* Added test for terminal worker

* Removed unused imports

* Removed datastore from confirm listener

* Removed datastore from confirm listener

* Removed storage ref from RabbitTaskInformation

* Minor refactoring

* added docker build section

* Update republishing flow

* Removed redundant testng.xml

* Simplified test setup

* Update WorkerQueueConsumerImpl.java

* Update WorkerConfirmListenerTest.java

---------

Co-authored-by: David Milligan <dmilligan@opentext.com>

andyreidz committed:
Update WorkerConfirmListenerTest.java

andyreidz committed:
Update WorkerConfirmListenerTest.java

andyreidz reviewed (approved):
Let's discuss at standup when to merge this.

dermot-hardy committed:
Code review: Some minor changes

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -44,76 +56,189 @@
+            try (final var inputStream = dataStore.retrieve(taskMessageStorageRef)) {
+                return inputStream.readAllBytes();
+            } catch (final IOException | DataStoreException e) {
+                final RabbitTaskInformation taskInformation = new RabbitTaskInformation(String.valueOf(inboundMessageId), true);
```
This message didn't crash the worker so why are we declaring it to be a poison message?

DavidGerald-Milligan replied:
I think this was to send it directly to the reject queue since the expected message was not in the data store.  @andyreidz  ?

andyreidz replied:
Yes if we fail to retrieve from the datastore we treat it as poison. 

In com.github.cafdataprocessing.workers.document.fieldvalues.ReferenceFieldValue we throw IOException if there's a problem retrieving from DataStore.

In Entity Extract com.hpe.caf.worker.entityextract.core.CachedInputData.initialize we catch IOException and throw RuntimeException recording it as a Failure on the document.

andyreidz replied:
Updated to check for a ReferenceNotFoundException which is the only non-transient type of error. Only this case is treated as a poison message.

In other cases disconnectIncoming is triggered.

Thoughts @dermot-hardy ?

dermot-hardy replied:
That sounds reasonable.  Later when we have enough of the workers updated we should test it in some real workflow scenarios to make sure it behaves as expected.

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -44,76 +56,189 @@
+            errorTaskInformation.incrementResponseCount(true);
+            final var publishHeaders = new HashMap<String, Object>();
+            publishHeaders.put(RabbitHeaders.RABBIT_HEADER_CAF_WORKER_REJECTED, REJECTED_REASON_TASKMESSAGE);
+            publishHeaders.put(RabbitHeaders.RABBIT_HEADER_CAF_PAYLOAD_OFFLOADING_STORAGE_REF, taskMessageStorageRefOpt.orElse(null));
```
Is setting this header to null safe?  Shouldn't it just not be set?

DavidGerald-Milligan replied:
https://github.com/WorkerFramework/worker-framework/pull/214/commits/0fd03112e24dec1914a26eac871fa38eed41fcd3

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -44,76 +56,189 @@
+        final boolean isPoison
+    ) {
+        final TrackingInfo trackingInfo = taskMessage.getTracking();
+        final String trackingJobTaskId = trackingInfo != null ? trackingInfo.getJobTaskId() : "untracked";
```
What does a normal job task id look like?  Would there be any confusion if someone created a job with "untracked" as the job id?  Presumably not because the partition is in there somewhere?

DavidGerald-Milligan replied:
The intent was to easily spot orphaned offloaded payloads in the datastore.  Would the confusion lie in job monitoring?

dermot-hardy replied:
What does the actual directory structure look like?

DavidGerald-Milligan replied:
Where this was tested on larry:
/etc/store/dataprocessing-update-entities-in/tenant-dave-001\:eca_g_503e01c3-e429-4b01-828c-90e04a8c058b.3.1.1/3e09ed9a-1c7d-4f71-8436-949f7566321d

where 3e09ed9a-1c7d-4f71-8436-949f7566321d is the offloaded payload

dermot-hardy replied:
I took a look on Larry to see what this looks like:
![image](https://github.com/user-attachments/assets/bf60e682-80ed-4b27-be59-0d9ec6e99c86)

It seems worrying that the queues are at the same directory level as the `batches` directory and the `device...` directories.  Shouldn't there be another directory level coming before this.

dermot-hardy replied:
Also looking inside the `dataprocessing-update-entities-in` folder I see the following &mdash; a directory for each task:

    ├── tenant-dave-001:eca_g_2b783514-9c3f-4e4d-a866-eb247ab590df.2.1.1
    ├── tenant-dave-001:eca_g_44fabc03-b023-49e2-a852-fd5d4ec4ed45.5.1.1
    ├── tenant-dave-001:eca_g_503e01c3-e429-4b01-828c-90e04a8c058b.3.1.1
    ├── tenant-dave-001:eca_g_61ff8eff-163a-4ca0-a82a-4de956e0b5c1.2.1.1
    └── tenant-dave-001:eca_g_682de301-7ba4-4dc6-8dc1-74d65c3c3f1f.1.1.1

Ultimately this will fail, because there will be too many directories under this one node.

It needs to be split out, ideally by tenant and then job id, so a directory structure more like this:

    └───tenant-dave-001
        ├───eca_g_2b783514-9c3f-4e4d-a866-eb247ab590df
        │   └───2.1.1
        ├───eca_g_44fabc03-b023-49e2-a852-fd5d4ec4ed45
        │   └───5.1.1
        ├───eca_g_503e01c3-e429-4b01-828c-90e04a8c058b
        │   └───3.1.1
        ├───eca_g_61ff8eff-163a-4ca0-a82a-4de956e0b5c1
        │   └───2.1.1
        └───eca_g_682de301-7ba4-4dc6-8dc1-74d65c3c3f1f
            └───1.1.1

andyreidz replied:
Let's introduce an additional configuration item to control the root directory to address Dermot's point about the queue directories being at the same level as where the staging service batches are.

I suggest using "queues" as the default value.

andyreidz replied:
There's something wrong with this.

<img width="245" alt="image" src="https://github.com/user-attachments/assets/d0177eb3-12e4-44b2-8691-c9cff58659ec" />

It looks like the rest of the queue name after » is missing.

DavidGerald-Milligan replied:
The dataprocessing-classification-in>> folder is not created by the offloading which was not enabled on worker-classification

DavidGerald-Milligan replied:
The queue name is listed in RabbitMQ as
![image](https://github.com/user-attachments/assets/f90c0d31-655c-494d-bf37-5774d1b2bd5e)

and the queue directory is created as
![image](https://github.com/user-attachments/assets/56ea8930-bd00-424d-ba0b-8af71e3fc2f0)

andyreidz replied:
Ahh, of course.

DavidGerald-Milligan replied:
The offloading directory has been updated as suggested

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -44,76 +56,189 @@
             taskInformation.incrementResponseCount(true);
-            publisherEventQueue.add(new WorkerPublishQueueEvent(delivery.getMessageData(), delivery.getEnvelope().getRoutingKey(),
-                    taskInformation));
+            publisherEventQueue.add(new WorkerPublishQueueEvent(taskMessageByteArray, routingKey, taskInformation, deliveryHeaders));
```
I notice the `deliveryHeaders` are being added here in this rejection circumstance?  Is that intentional?

DavidGerald-Milligan replied:
Yes this would include the offloading header which, if present, would still be required when processing the rejected message.

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -131,6 +256,17 @@ public void processAck(long tag)
+            return;
+        }
+
+        final String datastorePayloadReference = offloadedPayloads.get(tag);
```
I see we're deleting the payload from the datastore here, but we don't seem to ever be removing the entry from the `offloadedPayloads` map.  This seems like a memory leak.

DavidGerald-Milligan replied:
https://github.com/WorkerFramework/worker-framework/pull/214/commits/8e66c073e9d6b49adce98113a2976740cccf6c39

DavidGerald-Milligan committed:
Check for storage ref before setting header

DavidGerald-Milligan committed:
Removing offloadedPayload ref

DavidGerald-Milligan committed:
Updated offloading location

DavidGerald-Milligan committed:
removed unused imports

DavidGerald-Milligan committed:
corrected javadoc

andyreidz committed:
Handle transient exception

Only ReferenceNotFoundException is not transient.

andyreidz committed:
Disconnect

andyreidz committed:
Rework to store TaskData in DataStore

andyreidz commented:
@DavidGerald-Milligan Please review the latest commit, for testing we want to verify that transient and fatal datastore exceptions are handled correctly. Any associated job should be marked as encountering an error.

You can test fatal errors by deleting one of the offloaded payloads from the DataStore while running through a workflow. I haven't thought of an easy way to trigger a transient error, such as the DataStore being temporarily unavailable, let me know if you think of something.

andyreidz committed:
Formatting

andyreidz committed:
use TaskData length

andyreidz reviewed (approved):
@dermot-hardy would also like your review.

andyreidz committed:
Update WorkerApplication.java

dermot-hardy commented on worker-core/src/main/java/com/github/workerframework/core/WorkerCore.java:
```diff
@@ -151,12 +149,12 @@ public CoreTaskCallback(final Codec codec, final WorkerStats stats, final Worker
             Objects.requireNonNull(taskInformation);
             stats.incrementTasksReceived();
-            stats.getInputSizes().update(taskMessage.length);
+            stats.getInputSizes().update(taskMessage.getTaskData().length);
```
Are we sure that `.getTaskData()` never returns `null` at this point?

andyreidz committed:
We need the invalid task queue

andyreidz commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -34,86 +42,234 @@
+        } catch (final InvalidTaskException e) {
+            LOG.error("Cannot register new message, rejecting {}", inboundMessageId, e);
             taskInformation.incrementResponseCount(true);
-            publisherEventQueue.add(new WorkerPublishQueueEvent(delivery.getMessageData(), retryRoutingKey, taskInformation,
```
There is an existing fault here. 

When an InvalidTaskException is caught we're sending a new message to the retryRoutingKey, which is typically the input queue and we'd get stuck in an infinite loop.

Looking at how InvalidTaskException is handled elsewhere I'm now passing this queue name through to the WorkerQueueConsumerImpl

https://github.com/search?q=InvalidTaskQueue+caf&type=code&p=1
https://gitlab.otxlab.net/search?group_id=86711&scope=blobs&search=InvalidTaskQueue

**Document Workers have CAF_WORKER_FAILURE_QUEUE which is mapped to InvalidTaskQueue**

https://github.com/CAFDataProcessing/worker-document/blob/200253ea9fad190866e98b22119c19dd8eb0827a/worker-document/src/main/java/com/github/cafdataprocessing/workers/document/config/DocumentWorkerConfiguration.java#L76

https://github.com/CAFDataProcessing/worker-document/blob/200253ea9fad190866e98b22119c19dd8eb0827a/worker-document/src/main/java/com/github/cafdataprocessing/workers/document/DocumentWorkerAdapter.java#L108

**Production service is mapping OutputQueue to InvalidTaskQueue**

https://gitlab.otxlab.net/cybersecurity-enterprise/cddri/app/production/-/blob/master/production-service/src/main/java/com/hpe/verity/worker/production/ProductionServiceFactory.java

**Worker markup is using outputQueue**

https://github.com/CAFDataProcessing/worker-markup/blob/f5b2e1e6d4fc5f82b805ccdc670da8d30e3dc01d/worker-markup/src/main/java/com/github/cafdataprocessing/workers/markup/MarkupWorkerFactory.java#L84

**Worker batch is using outputQueue**

https://github.com/JobService/worker-batch/blob/f4da1bc2a089a76d7eefc1f13447348c99899fb1/worker-batch/src/main/java/com/github/jobservice/workers/batch/BatchWorkerFactory.java#L74

**Worker task stowing is using failureQueue**

https://github.com/JobService/worker-taskstowing/blob/7b80911ed0cd6fd8bc09bc741e0761f29a4f947d/worker-taskstowing/src/main/java/com/github/jobservice/workers/taskstowing/factory/TaskStowingWorkerFactory.java#L77

**Job service is using outputqueue**

https://github.com/JobService/job-service/blob/0d8f2942e7feb6de655e4d1601b18323155ccb7f/worker-jobtracking/src/main/java/com/github/jobservice/workers/jobtracking/JobTrackingWorkerFactory.java#L243

**Example of existing usage**

```
StreamingWorkerWrapper.run()

catch (InvalidTaskException e) {
            workerTask.setResponse(e);
```


`WorkerTaskImpl.setResponse(final InvalidTaskException invalidTaskException)`

```
        final TaskMessage invalidResponse = new TaskMessage(
            MoreObjects.firstNonNull(taskMessage.getTaskId(), ""),
            taskClassifier,
            taskMessage.getTaskApiVersion(),
            taskData,
            TaskStatus.INVALID_TASK,
            context,
**            workerFactory.getInvalidTaskQueue(), **
```

DavidGerald-Milligan replied:
Didn't we encounter this infinite loop early days? Was that caused by the null byte array deserialization?

DavidGerald-Milligan replied:
Actually I recall now, it was the null byte array when the header was missing, we fixed that in the message prioritization when we copied in the missing headers.

dermot-hardy commented on worker-api/src/main/java/com/github/workerframework/api/WorkerQueueProvider.java:
```diff
@@ -27,9 +28,11 @@ public interface WorkerQueueProvider
      * @throws QueueException if a WorkerQueue could not be created
      */
-    ManagedWorkerQueue getWorkerQueue(ConfigurationSource configurationSource, int maxTasks)
+    ManagedWorkerQueue getWorkerQueue(ConfigurationSource configurationSource, int maxTasks, String invalidQueue, ManagedDataStore dataStore, Codec codec)
```
Missing Javadoc param for `invalidQueue` and line greater than 138 characters long:

![image](https://github.com/user-attachments/assets/2c9b45b4-9508-41bd-85ea-6e7588be8cf9)

dermot-hardy commented on worker-core/src/main/java/com/github/workerframework/core/WorkerCore.java:
```diff
@@ -502,17 +487,16 @@ public void complete(final TaskInformation taskInformation, final String queue,
-                    workerQueue.publish(taskInformation, output, queue, Collections.emptyMap(), true);
-                    stats.getOutputSizes().update(output.length);
+                    workerQueue.publish(taskInformation, responseMessage, queue, Collections.emptyMap(), true);
+                    stats.getOutputSizes().update(responseMessage.getTaskData().length);
```
Same question here: Are we sure `.getTaskData()` will never be `null`?

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/InvalidDeliveryException.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Create a new InvalidDeliveryException with the specified message.
+     *
+     * @param message the message to include in the exception
```
Javadoc for `message` but not `messageId`.

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/InvalidDeliveryException.java:
```diff
@@ -0,0 +1,40 @@
+ */
+package com.github.workerframework.queues.rabbit;
+
+public class InvalidDeliveryException extends Exception {
```
Does this need to be `public`?

```suggestion
class InvalidDeliveryException extends Exception {
```

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/TransientDeliveryException.java:
```diff
@@ -0,0 +1,35 @@
+ */
+package com.github.workerframework.queues.rabbit;
+
+public class TransientDeliveryException extends Exception {
```
Does this need to be public?

```suggestion
class TransientDeliveryException extends Exception {
```

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/TransientDeliveryException.java:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * Create a new TransientDeliveryException with the specified message.
+     *
+     * @param message the message to include in the exception
```
Missing params in the Javadoc (or just remove the Javadoc if this class doesn't need to be public).

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitWorkerQueueConfiguration.java:
```diff
@@ -87,6 +87,23 @@ public class RabbitWorkerQueueConfiguration
+    /**
+     * The datastore directory to use for offloading payloads.
+     */
+    @NotNull
```
This could force worker updates to be required.  Can we not just accept `null` and have a default in the code for when that is the case?

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitWorkerQueue.java:
```diff
@@ -72,15 +81,27 @@ public final class RabbitWorkerQueue implements ManagedWorkerQueue
     {
         this.config = Objects.requireNonNull(config);
         this.maxTasks = maxTasks;
+        this.invalidQueue = invalidQueue;
```
The two other new parameters being added have `Objects.requireNonNull()` but this one doesn't.  Does that imply that we are ok with `null` being accepted here?

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitWorkerQueue.java:
```diff
@@ -140,11 +177,54 @@ public void publish(TaskInformation taskInformation, byte[] taskMessage, String
+        final StringBuilder path = new StringBuilder(Paths.get(config.getPayloadOffloadingDirectory(), routingKey).toString());
+        final Matcher matcher = JOB_TASK_ID_PATTERN.matcher(trackingJobTaskId);
+        if (matcher.find()) {
+            path.append("/" + matcher.group(1).replace(":", "/"));
```
```suggestion
            path.append('/').append(matcher.group(1).replace(":", "/"));
```

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitWorkerQueue.java:
```diff
@@ -140,11 +177,54 @@ public void publish(TaskInformation taskInformation, byte[] taskMessage, String
+        if (matcher.find()) {
+            path.append("/" + matcher.group(1).replace(":", "/"));
+            if (matcher.group(2) != null && !matcher.group(2).isEmpty()) {
+                path.append("/" + matcher.group(2));
```
```suggestion
                path.append('/').append(matcher.group(2));
```

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -34,86 +42,231 @@
         this.publisherEventQueue = Objects.requireNonNull(pubQueue);
         this.retryRoutingKey = Objects.requireNonNull(retryKey);
         this.retryLimit = retryLimit;
+        this.invalidRoutingKey = invalidKey;
+        this.dataStore = Objects.requireNonNull(dataStore);
+        this.codec = Objects.requireNonNull(codec);
+        this.disconnectCallback = disconnectCallback;
```
Using `Objects.requireNonNull()` with `dataStore` and `codec`, but not with `invalidKey` and `disconnectCallback` makes it look like `null` is acceptable for these, which I assume is not the case?

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -34,86 +42,231 @@
+            final var publishHeaders = new HashMap<String, Object>();
+
+            if(ex.getCause() != null && ex.getCause() instanceof ReferenceNotFoundException) {
+                publishHeaders.put(RabbitHeaders.RABBIT_HEADER_CAF_WORKER_REJECTED, REJECTED_REASON_PAYLOAD_OFFLOADING_TASKMESSAGE_DATASTOREE_REFERENCE_NOT_FOUND_ERROR);
```
```suggestion
                publishHeaders.put(RabbitHeaders.RABBIT_HEADER_CAF_WORKER_REJECTED,
                                   REJECTED_REASON_PAYLOAD_OFFLOADING_TASKMESSAGE_DATASTOREE_REFERENCE_NOT_FOUND_ERROR);
```

DavidGerald-Milligan commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -34,86 +42,231 @@
+    public static final String REJECTED_REASON_TASKMESSAGE_INVALID = "TASKMESSAGE_INVALID";
+    private static final String REJECTED_REASON_PAYLOAD_OFFLOADING_TASKMESSAGE_DATASTOREE_REFERENCE_NOT_FOUND_ERROR = 
+            "TASKMESSAGE_DATASTORE_REFERENCE_NOT_FOUND_ERROR";
+
```
typo? REJECTED_REASON_PAYLOAD_OFFLOADING_TASKMESSAGE_**DATASTOREE**_REFERENCE_NOT_FOUND_ERROR

DavidGerald-Milligan commented on worker-test/src/test/java/com/github/workerframework/workertest/WorkerTestBase.java:
```diff
@@ -51,8 +51,8 @@ public class WorkerTestBase {
     private static final String CAF_RABBITMQ_PASSWORD = "CAF_RABBITMQ_PASSWORD";
-    public static final String WEBDAV_URL = System.getProperty("WEBDAV_URL", "http://localhost:9090/webdav");
-    private static final Codec codec = new JsonCodec();
+    protected static final String WEBDAV_URL = System.getProperty("WEBDAV_URL", "http://localhost:9090/webdav");
```
I dont see any use of this outside WorkerTestBase

DavidGerald-Milligan commented on worker-test/src/test/java/com/github/workerframework/workertest/PayloadOffloadingIT.java:
```diff
@@ -79,6 +79,29 @@ public void checkOffloadedPayloadIsConsumedAndDeletedOnAck() throws Exception {
+
+            final TestWorkerQueueConsumer outboundConsumer = new TestWorkerQueueConsumer();
+            consume(channel, outboundConsumer, WORKER_OUT);
+            Assert.assertEquals("", new String(outboundConsumer.getLastDeliveredBody(), StandardCharsets.UTF_8));
```
Flip arguments, but shouldn't getLastDeliveredBody be returning the TaskMessage with an empty Taskdata byteArray?

DavidGerald-Milligan commented on worker-api/src/main/java/com/github/workerframework/api/TaskMessage.java:
```diff
@@ -53,7 +53,6 @@ public final class TaskMessage
      * The serialised data of the task-specific message.
      */
-    @NotNull
     private byte[] taskData;
```
The ctor has a null check: _this.taskData = Objects.requireNonNull(taskData);_

DavidGerald-Milligan commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -175,21 +339,43 @@ private void processReject(long id, boolean requeue)
+                    LOG.error("Failed to relocate offloaded payload for message id {} from {} to {}",
+                            inboundMessageId, deliveryQueue, retryRoutingKey, e);
+                    disconnectCallback.run();
+                }
```
Does retrieve only throw the ReferenceNotFoundException?
Don't we need to update the publish headers here too? And delete the now relocated data?

andyreidz committed:
Review feedback

Stop using rejected for invalid tasks to avoid confusion with a deprecated feature

andyreidz committed:
Configure reject queue for poison message tests

andyreidz committed:
Remove Objects.requireNonNull(taskData);

andyreidz commented:
@DavidGerald-Milligan Could you have another look at the recent changes?

DavidGerald-Milligan commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -166,10 +166,11 @@ public void processDelivery(Delivery delivery)
             LOG.warn("Transient error processing message id {}, disconnecting.", inboundMessageId, e);
             offloadedPayloadsToDelete.remove(inboundMessageId);
+            //Disconnect the channel to allow for a reconnect when the HealthCheck passes.
             disconnectCallback.run();
```
Only vaguely related but I see that the readme states  " _- If a transient error occurs in processing, throw a TaskRejectedException_" does that need updated?

andyreidz replied:
It's still correct in that when the worker is doing it work (in doWork) a TaskRejectedException will be thrown but this is before the worker is given the task so handled differently.

DavidGerald-Milligan commented on worker-test/src/test/java/com/github/workerframework/workertest/PayloadOffloadingIT.java:
```diff
@@ -161,9 +183,9 @@ public void checkOffloadedPayloadIsDeletedOnTerminalWorker() throws Exception {
-
+            final Channel channel = prepareChannel(connection)) {
+            createQueues(channel, TERMINAL_WORKER_IN, TERMINAL_WORKER_OUT);
+            
```
whitespace

andyreidz committed:
Whitespace fix

DavidGerald-Milligan committed:
Increaed potential wait for message delivery

DavidGerald-Milligan committed:
US1009117-Reporting (#227)

DavidGerald-Milligan committed:
Removed reset of taskdata

DavidGerald-Milligan committed:
Competed required fields in tracking report

DavidGerald-Milligan committed:
formatting

DavidGerald-Milligan committed:
Add report update status

DavidGerald-Milligan committed:
Merged main

DavidGerald-Milligan committed:
Formatting

DavidGerald-Milligan committed:
Resolved conflicts

DavidGerald-Milligan committed:
cfg update

DavidGerald-Milligan committed:
get worker name in WorkerApplication

DavidGerald-Milligan committed:
Removed duplicate dependency

DavidGerald-Milligan committed:
updated test

DavidGerald-Milligan committed:
removed dependency

DavidGerald-Milligan committed:
Use workercfg when available

DavidGerald-Milligan committed:
declare invlaid queue

DavidGerald-Milligan committed:
updated module names in pom

DavidGerald-Milligan committed:
formatting

dermot-hardy commented:
What were all the changes made here after the review approvals?  Were they trivial?

DavidGerald-Milligan commented:
> What were all the changes made here after the review approvals? Were they trivial?

The main update was the [reporting PR](https://github.com/WorkerFramework/worker-framework/pull/227) which was merged with this, subsequent minor updates were made following testing of that.  

The only other noteable update:
We had to [pass the workerName](https://github.com/WorkerFramework/worker-framework/pull/214/commits/d795cc7ac1233e525ed0154d38921fd554c6f606) to enable reporting, because an issue was discovered when integrating with workers that implemented an AbstractWorkerConfiguration that dis not have a WorkerConfiguration.

Anything else was trivial.

All changes were however discussed at length with, and approved by @andyreidz

andyreidz commented:
@dermot-hardy I'm not sure why the reviews from June weren't marked as stale by GitHub. David is right that the main change since the last review was to correct the reporting of invalid tasks to job service but you haven't seen those changes and I very much appreciate your input to them. Is it practical for you to comment on this merged PR and then we decide next step?

dermot-hardy commented on worker-core/pom.xml:
```diff
@@ -48,6 +48,10 @@
             <groupId>com.github.workerframework</groupId>
             <artifactId>worker-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.github.workerframework</groupId>
+            <artifactId>worker-caf</artifactId>
+        </dependency>
```
`worker-core` taking a dependency on `worker-caf` is very unexpected given that `worker-caf` is meant to be just some abstract classes to help implement workers.  What is the justification for this?

dermot-hardy commented on worker-core/src/main/java/com/github/workerframework/core/WorkerApplication.java:
```diff
@@ -370,4 +372,25 @@ private static HealthCheckConfiguration createHealthCheckConfiguration(
+     * sometimes implements WorkerConfiguration.  In those cases we take the workerName from
+     * the WorkerConfiguration, otherwise we take the workerName from AbstractWorkerFactory.
+     *
+     * Note there are occasions where the workerName is defined differently in AbstractWorkerFactory
+     * to that in WorkerConfiguration. e.g. worker-batch.
```
This seems to be the answer to my previous question but this seems like a bad idea.  I really don't like `worker-core` having special logic for `AbstractWorkerFactory`, which is just a barely used (badly written) convenience class.  The `AbstractWorkerFactory.getWorkerName()` is just badly named &mdash; it's not even the name of the worker.

dermot-hardy commented on worker-caf/src/main/java/com/github/workerframework/caf/AbstractWorkerFactory.java:
```diff
@@ -159,7 +159,7 @@ private T verifyWorkerTask(
      * @return the name of the worker to use when checking the task type
      */
-    protected abstract String getWorkerName();
+    public abstract String getWorkerName();
```
I don't like this change.  Discussed in previous comments.

dermot-hardy commented on worker-queue-rabbit/pom.xml:
```diff
@@ -48,6 +48,10 @@
             <groupId>com.github.workerframework</groupId>
             <artifactId>worker-configs</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.github.workerframework</groupId>
+            <artifactId>worker-tracking-report</artifactId>
+        </dependency>
```
Why does `worker-queue-rabbit` now have a dependency on `worker-tracking-report`?  These seem like they should be very unrelated things.

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/RabbitWorkerQueue.java:
```diff
@@ -107,13 +131,31 @@ public void start(TaskCallback callback)
             publisher = new EventPoller<>(2, publisherQueue, publisherImpl);
             declareWorkerQueue(incomingChannel, config.getInputQueue());
             declareWorkerQueue(outgoingChannel, config.getRetryQueue());
+            declareWorkerQueue(outgoingChannel, invalidQueue);
```
Does this mean that we're creating lots of invalid queues that may never be used?  (I'm not sure if that's an issue if it does mean that.  It might not be.)

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -63,57 +104,249 @@ public WorkerQueueConsumerImpl(TaskCallback callback, RabbitMetricsReporter metr
+            final var taskData = inputStream.readAllBytes();
+            offloadedPayloadsToDelete.put(inboundMessageId, taskMessageStorageRef);
+            return taskData;
+        } catch (final IOException | DataStoreException ex) {
+            if (ex instanceof ReferenceNotFoundException) {
+                throw (ReferenceNotFoundException)ex;
+            }
```
Why is the `ReferenceNotFoundException` not just being caught first?

```suggestion
        } catch (final ReferenceNotFoundException ex) {
            throw ex;
        } catch (final IOException | DataStoreException ex) {
```

dermot-hardy commented on worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java:
```diff
@@ -63,57 +104,249 @@ public WorkerQueueConsumerImpl(TaskCallback callback, RabbitMetricsReporter metr
+                    sendFailureTrackingReport(taskMessage, exceptionMesssage, taskInformation);
+                }
+            }
+        } catch (CodecException e) {
```
```suggestion
        } catch (final CodecException e) {
```
